File: LP-3/DAA/2HuffmanEncoding.cpp
#include <bits/stdc++.h>;
using namespace std;

struct MinHeapNode
{
    char d;
    unsigned frequency;
    MinHeapNode *lChild, *rChild;

    MinHeapNode(char d, unsigned frequency)
    {
        lChild = rChild = NULL;
        this-&gt;d = d;
        this-&gt;frequency = frequency;
    }
};

//function to compare
struct compare
{
    bool operator()(MinHeapNode *l, MinHeapNode *r)
    {
        return (l-&gt;frequency &gt; r-&gt;frequency);
    }
};

void printCodes(struct MinHeapNode *root, string str)
{
    if (!root)
        return;

    if (root-&gt;d != &#39;$&#39;)
        cout &lt;&lt; root-&gt;d &lt;&lt; &quot;: &quot; &lt;&lt; str &lt;&lt; &quot;\n&quot;;

        printCodes(root-&gt;lChild, str + &quot;0&quot;);
        printCodes(root-&gt;rChild, str + &quot;1&quot;);
}

void showpq(
        priority_queue&lt;MinHeapNode *, vector&lt;MinHeapNode *&gt;, compare &gt; g)
    {
        while (!g.empty()) {
            cout &lt;&lt; &#39; &#39; &lt;&lt; g.top()-&gt;d;
            g.pop();
    }
        cout<<"\n";
    }

void HuffmanCodes(char d[], int frequency[], int size)
    {
        struct MinHeapNode *lChild, *rChild, *top;

        priority_queue&lt;MinHeapNode *, vector&lt;MinHeapNode *&gt;, compare&gt; minHeap;

        for (int i = 0; i &lt; size; ++i)
            minHeap.push(new MinHeapNode(d[i], frequency[i]));

        while (minHeap.size() != 1)
        {
            lChild = minHeap.top();
            minHeap.pop();

            rChild = minHeap.top();
            minHeap.pop();

            top = new MinHeapNode(&#39;$&#39;, lChild-&gt;frequency + rChild-&gt;frequency);

            top-&gt;lChild = lChild;
            top-&gt;rChild = rChild;

            minHeap.push(top);

        }
        printCodes(minHeap.top(), &quot;&quot;);
    }


int main()
{

    char arr[] = {'a','b','c','d','e','f'};
    int frequency[] = {5, 9, 12, 13, 16, 45};
    int size = sizeof(arr) / sizeof(arr[0]);
    HuffmanCodes(arr, frequency, size);
    return 0;
}

File: LP-3/DAA/2. Huffman encoding.py
# Write a program to implement Huffman Encoding using a greedy strategy.

import collections, heapq

# d - tree direction (0/1)
Node = collections.namedtuple('Node',['d','freq','lchild','rchild'])

def print_codes(root, code):
    if root is not None:
        if(root.d != "$"):
            print(f"{root.d}: {code}")
        print_codes(root.lchild, code + "0") 
        print_codes(root.rchild, code + "1")

def HuffmanCodes(data, frequency):
    min_heap = []
    for i in range(len(data)):
        heapq.heappush(min_heap, Node(data[i], frequency[i], None, None))
    while(len(min_heap) > 1):
        lchild = heapq.heappop(min_heap)
        rchild = heapq.heappop(min_heap)
        top = Node("$", lchild.freq + rchild.freq, lchild, rchild)
        heapq.heappush(min_heap, top)
    print("Huffman Code: ")
    print_codes(min_heap[0], "") 

def main():
    # # characters for huffman tree & # frequency of characters
    arr, freq = ['a', 'b', 'c', 'd', 'e', 'f'], [5, 9, 12, 13, 16, 45]

    HuffmanCodes(arr, freq)

main()